dll/win32/shell32/CShellDispatch.cpp

```diff
@@ -343,7 +468,13 @@ HRESULT STDMETHODCALLTYPE CShellDispatch::IsServiceRunning(BSTR name, VARIANT *r
 HRESULT STDMETHODCALLTYPE CShellDispatch::CanStartStopService(BSTR service, VARIANT *ret)
 {
     TRACE("(%p, %ls, %p)\n", this, service, ret);
-    return E_NOTIMPL;
+
+    SC_HANDLE hSvc = OpenServiceHelper(service, SERVICE_START | SERVICE_STOP);
+    if (hSvc)
+        CloseServiceHandle(hSvc);
+    V_VT(ret) = VT_BOOL;
+    V_BOOL(ret) = (hSvc ? VARIANT_TRUE : VARIANT_FALSE);
```

Comment:
Use after free here. You're already closing the `hSvc` service handle at line 474, in case it's valid. After closing, it should become invalid (`NULL` or equal to `INVALID_HANDLE_VALUE`). Then you're checking whether it is valid at line 476, and basing on the result condition, you're setting the resulting value. Since the handle is always `NULL` in this case, this condition will always be set to `FALSE`.
It's better to do as follows:
```suggestion
    V_VT(ret) = VT_BOOL;
    V_BOOL(ret) = (hSvc ? VARIANT_TRUE : VARIANT_FALSE);
    if (hSvc)
        CloseServiceHandle(hSvc);
```

Reply:
@oleg-dubinskiy No it's not a use-after-free per se. The value of stored in the `hSvc` variable won't change by the CloseServiceHandle, i.e. If there was a non-NULL value stored in it (via a successful call to OpenService) it still is stored that way even after the CloseServiceHandle call. And so we can still use it to check whether the OpenService call succeeded because a service existed. (Vice-versa: if it was NULL because OpenService failed, it's still NULL.)

Reply:
Oh, indeed. `CloseServiceHandle` doesn't actually close/invalidate the handle, MSDN also confirms your statement. My mistake @HBelusca. :sweat_smile: I though it does the similar thing to `CloseHandle` for services...

Reply:
Yes it does a similar thing as `CloseHandle` : It dereferences the underlying object the handle describes, closing the object if its reference reaches zero, and invalidates the handle i.e. makes its value _**not**_ refer to the object anymore. _**BUT**_, the handle variable value doesn't change.